### DISCUSSION TITLE
What causes 'Error: Another active Homebrew update process is already in progress.' repeatedly?

### DISCUSSION THREAD

**Discussion by netopsengineer** [Homebrew/discussions#5478] [Getting started] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.3.9
ORIGIN: https://github.com/Homebrew/brew
HEAD: 1c98a33a376e5b35ebc801d0f7d6ba7b9eb118ac
Last commit: 25 hours ago
Core tap JSON: 09 Jul 02:10 UTC
Core cask tap JSON: 09 Jul 02:10 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 16
Homebrew Ruby: 3.3.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/3.3.3/bin/ruby
CPU: 16-core 64-bit kabylake
Clang: 15.0.0 build 1500
Git: 2.45.2 => /usr/local/bin/git
Curl: 8.6.0 => /usr/bin/curl
macOS: 14.5-x86_64
CLT: 15.3.0.0.1.1708646388
Xcode: N/A
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

I can't track down the behavior that is leading to this condition, I used to be able to `brew update && brew upgrade` but recently that seems to fail every time with the error listed in the title, and I must go delete the lock file.

When I `ps aux | grep brew` when its locked nothing shows up.  I thought it was the `&&` only, but I have seen where the `brew` raycast extension also seemed to cause it, so I can't be certain what leads to it at this point.

Can anyone recommend some troubleshooting steps that might uncover what led to the lock, my assumption is that brew is crashing in the background perhaps, and it is what created and left the lock.

**Comment by konnano** [comment, marked as answer]
I got the same symptoms.
When I checked with the top -u command, git was running in the background.
I waited for a while, and when git was terminated, it was cured.

**Comment by p-linnane** [comment]
As @konnano advised, it is most likely a hung git process and should resolve itself once you kill the process.

**Reply by netopsengineer** [reply]
Appreciate you both chiming in, that would certainly be the commonality between different entry points leading to the same condition.  This only seems to have started the past few months and been a brew user for many happy years.

**Comment by netopsengineer** [comment]
Wanted to post a follow up:

One thing that I believe could have been causing my issue, was this particular `git` setting I had enabled.  It is a file watcher daemon that tries to create better performance on large repos, and it does do that, but seems as a side affect, brew thinks something is still going on with the git process once its done and holds the lock.

`git config --global core.fsmonitor true`   I flipped it to false, and so far no issues.  I have used this setup for quite some time with it enabled, so it wasn't always troublesome, but seems to be the culprit right now.

**Reply by p-linnane** [reply]
See also https://github.com/Homebrew/brew/pull/17794

**Reply by konnano** [reply]
Happened on the same MacBook
OS 10.13 showed an error message and the update was aborted
Debian 12 only showed an error message and did not abort update

```
OS 10.13

(Desktop):git config -l
credential.helper=osxkeychain
filter.lfs.required=true
filter.lfs.clean=git-lfs clean -- %f
filter.lfs.smudge=git-lfs smudge -- %f
filter.lfs.process=git-lfs filter-process
user.name=konnano
user.email=73874687+konnano@users.noreply.github.com
safe.directory=/usr/local/Homebrew
(Desktop):brew config
HOMEBREW_VERSION: 4.3.9-371-g8e8ce4a-dirty
ORIGIN: https://github.com/Homebrew/brew
HEAD: 8e8ce4a2f7fb6044c83101ffc12c57d1b15938e1
Last commit: 16 hours ago
Core tap HEAD: efb1e0b0083a5fc828153a79f6c0dfa0c4418a22
Core tap last commit: 24 hours ago
Core tap JSON: 24 Jul 10:01 UTC
Core cask tap HEAD: 526b190184dc46fdf7a0eee5267a95f784899c8a
Core cask tap last commit: 25 hours ago
Core cask tap JSON: 24 Jul 10:01 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: :0
HOMEBREW_EDITOR: vim
HOMEBREW_MAKE_JOBS: 4
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.3.4 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/3.3.4/bin/ruby
CPU: quad-core 64-bit ivybridge
Clang: 10.0.0 build 1000
Git: 2.45.2 => /usr/local/bin/git
Curl: 7.54.0 => /usr/bin/curl
macOS: 10.13.6-x86_64
CLT: 10.1.0.0.1.1539992718
Xcode: 10.1
```  

```
Debian 12

(Desktop):git config -l
(Desktop):brew config
HOMEBREW_VERSION: 4.3.9-371-g8e8ce4a
ORIGIN: https://github.com/Homebrew/brew
HEAD: 8e8ce4a2f7fb6044c83101ffc12c57d1b15938e1
Last commit: 15 hours ago
Core tap HEAD: c71b82e70e45b6b316137ed3ed49158566583210
Core tap last commit: 2 hours ago
Core tap JSON: 24 Jul 10:01 UTC
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: :0.0
HOMEBREW_MAKE_JOBS: 2
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.3.4 => /home/linuxbrew/.linuxbrew/Cellar/ruby/3.3.4/bin/ruby
CPU: dual-core 64-bit ivybridge
Clang: 18.1.8
Git: 2.39.2 => /bin/git
Curl: 7.88.1 => /bin/curl
Kernel: Linux 6.1.0-18-amd64 x86_64 GNU/Linux
OS: Debian GNU/Linux 12 (bookworm) (bookworm)
Host glibc: 2.36
/usr/bin/gcc: 12.2.0
/usr/bin/ruby: N/A
glibc: 2.35_1
gcc@11: N/A
gcc: 14.1.0_2
xorg: N/A
```

**Comment by garthk** [comment]
To disable fsmonitor for brew taps in particular:

```sh
( prefix=$(brew --prefix)
  echo "$prefix/Homebrew"
  brew tap | while IFS= read -r tap; do echo "$prefix/Homebrew/Library/Taps/$tap"; done ) |
  while IFS= read -r repo; do git -C "$repo" config core.fsmonitor false; done
```

Tested on zsh and bash. Should be safe vs spaces in directory names.

That said, I reckon Brew should configure such itself.

**Reply by p-linnane** [reply]
Go ahead and open a PR and we'll be happy to review it.
